### PR TITLE
[zero] use double buffers to handle grad

### PR DESCRIPTION
**Pull request by ver217**
In order to support partial offload, we use double buffers to handle grad (`fp16_grad` and `fp32_grad`). `fp16_grad` is always on GPU, and `fp32_grad` may be partially on GPU and CPU.

### CODE REVIEW

File: colossalai/zero/sharded_model/sharded_model_v2.py [hunk 1]
```diff
@@ -216,23 +224,7 @@ def _reduce_scatter_callback(self, param: Parameter, reduced_grad: torch.Tensor)
...
-
-        if param.col_attr.grad is None:
-            param.col_attr.grad = reduced_grad.data
-        else:
-            # When dp size = 1
-            # param.col_attr.grad is local accumulated grad shard (full but flatten)
-            # But reduced_grad here is full grad
-            # We should call `view_as`
-            param.col_attr.grad.add_(reduced_grad.data.view_as(param.col_attr.grad))
+        param.col_attr.fp16_grad = reduced_grad.data
```
Comment (feifeibear) [inline colossalai/zero/sharded_model/sharded_model_v2.py:227]: do you need to remove param.grad?

**Comment by feifeibear** [comment]
Did you correctly manage the lifecycle of grad fp16? Because it can be released after it is copied to grad fp32.

File: colossalai/zero/sharded_model/sharded_model_v2.py (hunk 1 above)
Comment (ver217) [reply colossalai/zero/sharded_model/sharded_model_v2.py:227]: Yes. But we cannot set `param.grad` to `None` in parameters' hook. I tried to allocate an empty tensor and then free its storage before returning it, but unit tests failed. I'm debugging this problem.

**Comment by ver217** [comment]
> Did you correctly manage the lifecycle of grad fp16? Because it can be released after it is copied to grad fp32.

Yes. It's set to `None` after being copied.

**Comment by feifeibear** [comment]
Now the adam use grad fp32 to update? Did you implement it?

**Comment by ver217** [comment]
> Now the adam use grad fp32 to update? Did you implement it?

Yes. It always uses fp32 grad.

File: colossalai/zero/sharded_model/sharded_model_v2.py (hunk 1 above)
Comment (feifeibear) [reply colossalai/zero/sharded_model/sharded_model_v2.py:227]: set to None is enough

**Review by feifeibear — APPROVED** [review]

**Pull request merged**